uniref-rest/src/test/java/org/uniprot/api/uniref/queue/AsyncDownloadIntegrationTest.java

```diff
@@ -0,0 +1,326 @@
+        doThrow(new MessageListenerException("Forcing to throw unexpected exception"))
+                .when(this.uniRefMessageListener)
+                .dummyMethodForTesting(jobId, JobStatus.ERROR);
+        this.messageService.sendMessage(request);
+        verify(this.messageService, never()).alreadyProcessed(jobId);
+        await().until(jobCreatedInRedis(jobId));
+        await().atMost(Duration.ofSeconds(20)).until(jobErrored(jobId));
+        await().until(getJobRetriedCount(jobId), Matchers.equalTo(2));
+        await().until(getMessageCountInQueue(this.rejectedQueue), equalTo(1));
+        verify(this.uniRefMessageListener, atLeast(3)).onMessage(any());
+        verifyRedisEntry(query, jobId, List.of(JobStatus.ERROR), 2, true);
+        verifyIdsAndResultFilesDoNotExist(jobId);
+    }
+
+    private UniRefDownloadRequest createDownloadRequest(String query, MediaType format) {
+        UniRefDownloadRequest request = new UniRefDownloadRequest();
+        request.setQuery(query);
+        request.setFormat(format.toString());
+        request.setLargeSolrStreamRestricted(false);
+        return request;
+    }
+
+    private void verifyRedisEntry(
+            String query, String jobId, List<JobStatus> statuses, int retryCount, boolean isError) {
+        Optional<DownloadJob> optDownloadJob = this.downloadJobRepository.findById(jobId);
+        assertTrue(optDownloadJob.isPresent());
+        assertEquals(jobId, optDownloadJob.get().getId());
+        assertEquals(query, optDownloadJob.get().getQuery());
+        assertAll(
+                () -> assertNull(optDownloadJob.get().getSort()),
+                () -> assertNull(optDownloadJob.get().getFields()));
+        assertEquals(retryCount, optDownloadJob.get().getRetried());
+        assertTrue(statuses.contains(optDownloadJob.get().getStatus()));
+        assertEquals(isError, Objects.nonNull(optDownloadJob.get().getError()));
+        if (optDownloadJob.get().getStatus() == JobStatus.FINISHED) {
+            assertNotNull(optDownloadJob.get().getResultFile());
+            assertEquals(jobId, optDownloadJob.get().getResultFile());
+        } else {
+            assertNull(optDownloadJob.get().getResultFile());
+        }
+        assertAll(
+                () -> assertNotNull(optDownloadJob.get().getCreated()),
+                () -> assertNotNull(optDownloadJob.get().getUpdated()));
+    }
+
+    private void verifyMessageListener(int timesOnMessage, int timesAddHeader, int timesStreamIds) {
+        verify(this.uniRefMessageListener, atLeast(timesOnMessage)).onMessage(any());
+        verify(this.uniRefMessageListener, atLeast(timesAddHeader))
+                .addAdditionalHeaders(any(), any());
+        verify(this.uniRefMessageListener, atLeast(timesStreamIds)).streamIds(any());
+    }
+
+    private Callable<Boolean> jobCreatedInRedis(String jobId) {
+        return () -> this.downloadJobRepository.existsById(jobId);
+    }
+
```

Comment:
Move redis related helper methods to common class and use them in both uniprotkb and uniref's AsyncDownloadIntegration. 
Lets rename these two classes also. e.g. UniProtKBAsyncDownloadIntegration and UniRefAsyncDownloadIntegration